src/amici.cpp

```diff
@@ -2568,24 +2588,116 @@ ReturnData *getSimulationResults(UserData *udata, ExpData *edata, int *pstatus)
+
 freturn:
     storeJacobianAndDerivativeInReturnData(udata, tdata, rdata);
     freeTempDataAmiMem(udata, tdata, ami_mem, setupBdone, *pstatus);
+
+    if(originalParams) {
+        memcpy(p, originalParams, sizeof(double) * np);
+        free(originalParams);
+    }
     return rdata;
 }
 #endif
+
+
+#ifdef AMICI_WITHOUT_MATLAB
+void unscaleParameters(UserData *udata) {
+    switch(udata->am_pscale) {
+    case AMI_SCALING_LOG10:
+        for(int i = 0; i < np; ++i) {
+            p[i] = pow(10, p[i]);
+        }
+        break;
+    case AMI_SCALING_LN:
+        for(int i = 0; i < np; ++i)
+            p[i] = exp(p[i]);
+        break;
+    }
+}
+
+void applyChainRuleFactorToSimulationResults(const UserData *udata, ReturnData *rdata)
+{
+    if(udata->am_pscale == AMI_SCALING_NONE)
+        return;
+
+    // chain-rule factor: multiplier for am_p
+    double coefficient;
+
+    switch(udata->am_pscale) {
+    case AMI_SCALING_LOG10:
+        coefficient = log(10);
+        break;
+    case AMI_SCALING_LN:
+        coefficient = 1.0;
+        break;
+    }
+
+    if(sensi > 0) {
+        if(rdata->am_sllhdata)
+            for(int ip = 0; ip < np; ++ip)
+                sllhdata[ip] *= p[ip] * coefficient;
+
+        if(rdata->am_sxdata)
+            for(int ip = 0; ip < np; ++ip)
+                for(int ix = 0; ix < nx; ++ix)
+                    for(int it = 0; it < nt; ++it)
+                        sxdata[(ip*nx + ix)*nt + it] *= p[ip] * coefficient;
+
+        if(rdata->am_sydata)
+            for(int ip = 0; ip < np; ++ip)
+                for(int iy = 0; iy < ny; ++iy)
+                    for(int it = 0; it < nt; ++it)
+                        sydata[(ip*ny + iy)*nt + it] *= p[ip] * coefficient;
+
+        if(rdata->am_szdata)
+            for(int ip = 0; ip < np; ++ip)
+                for(int iz = 0; iz < nz; ++iz)
+                    for(int ie = 0; ie < nmaxevent; ++ie)
+                        szdata[(ip*nz + iz)*nmaxevent + ie] *= p[ip] * coefficient;
+
+        if(rdata->am_ssigmaydata)
+            for(int ip = 0; ip < np; ++ip)
+                for(int iy = 0; iy < ny; ++iy)
+                    for(int it = 0; it < nt; ++it)
+                        ssigmaydata[(ip*ny + iy)*nt + it] *= p[ip] * coefficient;
+
+        if(rdata->am_ssigmazdata)
+            for(int ip = 0; ip < np; ++ip)
+                for(int iz = 0; iz < nz; ++iz)
+                    for(int ie = 0; ie < nmaxevent; ++ie)
+                        ssigmazdata[(ip*nz + iz)*nmaxevent + ie] *= p[ip] * coefficient;
+    }
+
+    if(sensi_meth == AMI_SS) {
+        if(rdata->am_dxdotdpdata)
+            for(int ip = 0; ip < np; ++ip)
+                for(int ix = 0; ix < nx; ++ix)
+                    dxdotdpdata[ip*nx + ix] *= p[ip] * coefficient;
+
+        if(rdata->am_dydpdata)
+            for(int ip = 0; ip < np; ++ip)
+                for(int iy = 0; iy < ny; ++iy)
+                    dydpdata[ip*nx + iy] *= p[ip] * coefficient;
```

Comment:
this does not cover second order sensitivities, also I would like to not have this only for #ifdef AMICI_WITHOUT_MATLAB but also with matlab (should be relatively straightforward, just remove corresponding functionality in generateM.m, this would also serve as a sanity check whether all cases are covered)